pkg/cloudprovider/providers/openstack/openstack.go

```diff
@@ -524,3 +528,89 @@ func (os *OpenStack) Routes() (cloudprovider.Routes, bool) {
 
 	return r, true
 }
+
+// Implementation of sort interface for blockstorage version probing
+type APIVersionsByID []apiversions_v1.APIVersion
+
+func (apiVersions APIVersionsByID) Len() int {
+	return len(apiVersions)
+}
+
+func (apiVersions APIVersionsByID) Swap(i, j int) {
+	apiVersions[i], apiVersions[j] = apiVersions[j], apiVersions[i]
+}
+
+func (apiVersions APIVersionsByID) Less(i, j int) bool {
+	return apiVersions[i].ID > apiVersions[j].ID
+}
+
+func (os *OpenStack) Volumes() (IVolumes, error) {
+	switch os.bsOpts.BSVersion {
+	case "v1":
+		sClient, err := openstack.NewBlockStorageV1(os.provider, gophercloud.EndpointOpts{
+			Region: os.region,
+		})
+		if err != nil || sClient == nil {
+			glog.Errorf("Unable to initialize cinder client for region: %s", os.region)
+			return nil, err
+		}
+		return (IVolumes)(&VolumesV1{sClient, os.bsOpts}), nil
+	case "v2":
+		sClient, err := openstack.NewBlockStorageV2(os.provider, gophercloud.EndpointOpts{
+			Region: os.region,
+		})
+		if err != nil || sClient == nil {
+			glog.Errorf("Unable to initialize cinder v2 client for region: %s", os.region)
+			return nil, err
+		}
+		return (IVolumes)(&VolumesV2{sClient, os.bsOpts}), nil
+	case "auto":
+		sClient, err := openstack.NewBlockStorageV1(os.provider, gophercloud.EndpointOpts{
+			Region: os.region,
+		})
+		if err != nil || sClient == nil {
+			glog.Errorf("Unable to initialize cinder client for region: %s", os.region)
+			return nil, err
+		}
+
+		apiversions_v1.List(sClient).EachPage(func(page pagination.Page) (bool, error) {
+			apiversions, err := apiversions_v1.ExtractAPIVersions(page)
+			if err != nil {
+				glog.Errorf("Unable to list blockstorage API versions for region: %s", os.region)
+				return false, err
+			}
+
+			sort.Sort(APIVersionsByID(apiversions))
+			// take first "CURRENT" version
+			for _, version := range apiversions {
+				if version.Status == "CURRENT" {
+					switch version.ID {
+					case "v2.0":
+						os.bsOpts.BSVersion = "v2"
+						break
+					case "v1.0":
+						os.bsOpts.BSVersion = "v1"
+						break
+					}
+				}
+			}
+			return true, nil
+		})
+		// Autoprobing failed, no supported versions available
+		if os.bsOpts.BSVersion == "auto" {
```

Comment:
It is not, it will be (if probe is successful) switched to highest supported and available version on lines 589/592,
If it is left in "auto" that means that probe failed and it errors out with a corresponding error message.